### PR TITLE
fuse,meta: support GETFLAGS/SETFLAGS ioctl

### PR DESCRIPTION
**Pull request by qingyunha**

### CODE REVIEW

**Comment by qingyunha** [comment]
close #775

Hi, I have a init implement for support ioctl. It's not complete and I hope your help, see if this is the right direction

**Comment by davies** [comment]
@qingyunha This looks pretty good, some details could be tuned.

File: pkg/meta/base.go [hunk 1]
```diff
@@ -740,6 +740,13 @@ func (m *baseMeta) Link(ctx Context, inode, parent Ino, name string, attr *Attr)
 	if name == "" {
 		return syscall.ENOENT
 	}
+	err := m.GetAttr(ctx, inode, attr)
+	if err != 0 {
+		return err
+	}
+	if (attr.Flags&FlagAppend) != 0 || (attr.Flags&FlagImmutable) != 0 {
```
Comment (davies) [inline pkg/meta/base.go:747]: Check this inside the transaction to avoid race and the `GetAttr()`

Comment (qingyunha) [reply pkg/meta/base.go]: done

**Review by zhijian-pro — COMMENTED** [review]
`doMknod` should check parent dir's flags.
```
if (pattr.Flags & FlagImmutable) != 0 {
    return syscall.EPERM
}
```

File: pkg/meta/base_test.go [hunk 2]
```diff
@@ -1426,3 +1427,54 @@ func testConcurrentDir(t *testing.T, m Meta) {
...
+		t.Fatalf("setattr d: %s", st)
+	}
+	if st := m.Create(ctx, d, "f", 0644, 022, 0, &inode, attr); st != 0 {
+		t.Fatalf("create f: %s", st)
+	}
+	if st := m.Unlink(ctx, d, "f"); st != syscall.EPERM {
+		t.Fatalf("unlink f : %s", st)
+	}
+	attr.Flags = FlagAppend | FlagImmutable
+	if st := m.SetAttr(ctx, inode, SetAttrFlag, 0, attr); st != 0 {
```
Comment (zhijian-pro) [inline pkg/meta/base_test.go:1474]:
```suggestion
	if st := m.SetAttr(ctx, d, SetAttrFlag, 0, attr); st != 0 {
```

File: pkg/meta/base_test.go [hunk 3]
```diff
@@ -1426,3 +1427,54 @@ func testConcurrentDir(t *testing.T, m Meta) {
...
+		t.Fatalf("setattr d: %s", st)
+	}
+	if st := m.Create(ctx, d, "f2", 0644, 022, 0, &inode, attr); st != syscall.EPERM {
+		t.Fatalf("create f2: %s", st)
+	}
```
Comment (zhijian-pro) [inline pkg/meta/base_test.go:1479]: add more test
```
        var Immutable Ino
	if st := m.Mkdir(ctx, 1, "ImmutFile", 0640, 022, 0, &Immutable, attr); st != 0 {
		t.Fatalf("mkdir d: %s", st)
	}
	attr.Flags = FlagImmutable
	if st := m.SetAttr(ctx, Immutable, SetAttrFlag, 0, attr); st != 0 {
		t.Fatalf("setattr d: %s", st)
	}
	if st := m.Create(ctx, Immutable, "f2", 0644, 022, 0, &inode, attr); st != syscall.EPERM {
		t.Fatalf("create f2: %s", st)
	}

	var src1, dst1, mfile Ino
	attr.Flags = 0
	if st := m.Mkdir(ctx, 1, "src1", 0640, 022, 0, &src1, attr); st != 0 {
		t.Fatalf("mkdir src1: %s", st)
	}
	if st := m.Create(ctx, src1, "mfile", 0644, 022, 0, &mfile, attr); st != 0 {
		t.Fatalf("create mfile: %s", st)
	}
	if st := m.Mkdir(ctx, 1, "dst1", 0640, 022, 0, &dst1, attr); st != 0 {
		t.Fatalf("mkdir dst1: %s", st)
	}

	attr.Flags = FlagAppend
	if st := m.SetAttr(ctx, src1, SetAttrFlag, 0, attr); st != 0 {
		t.Fatalf("setattr d: %s", st)
	}
	if st := m.Rename(ctx, src1, "mfile", dst1, "mfile", 0, &mfile, attr); st != syscall.EPERM {
		t.Fatalf("rename d: %s", st)
	}

	attr.Flags = FlagImmutable
	if st := m.SetAttr(ctx, src1, SetAttrFlag, 0, attr); st != 0 {
		t.Fatalf("setattr d: %s", st)
	}
	if st := m.Rename(ctx, src1, "mfile", dst1, "mfile", 0, &mfile, attr); st != syscall.EPERM {
		t.Fatalf("rename d: %s", st)
	}

	if st := m.SetAttr(ctx, dst1, SetAttrFlag, 0, attr); st != 0 {
		t.Fatalf("setattr d: %s", st)
	}
	if st := m.Rename(ctx, src1, "mfile", dst1, "mfile", 0, &mfile, attr); st != syscall.EPERM {
		t.Fatalf("rename d: %s", st)
	}

	var delFile Ino
	if st := m.Create(ctx, 1, "delfile", 0644, 022, 0, &delFile, attr); st != 0 {
		t.Fatalf("create f: %s", st)
	}
	attr.Flags = FlagImmutable | FlagAppend
	if st := m.SetAttr(ctx, delFile, SetAttrFlag, 0, attr); st != 0 {
		t.Fatalf("setattr d: %s", st)
	}
	if st := m.Unlink(ctx, 1, "delfile"); st != syscall.EPERM {
		t.Fatalf("unlink f : %s", st)
	}
```

File: pkg/meta/redis.go [hunk 4]
```diff
@@ -1224,6 +1228,9 @@ func (m *redisMeta) doUnlink(ctx Context, parent Ino, name string) syscall.Errno
 		if pattr.Typ != TypeDirectory {
 			return syscall.ENOTDIR
 		}
+		if (pattr.Flags&FlagAppend) != 0 || (pattr.Flags&FlagImmutable) != 0 {
+			return syscall.EPERM
+		}
```
Comment (zhijian-pro) [inline pkg/meta/redis.go:1233]: should check itself flags in [redis.go#L1249](https://github.com/juicedata/juicefs/blob/b3c8473d9f7dd308798f86d39d211b13d8b5d6c6/pkg/meta/redis.go#L1249)
 
```go
if (attr.Flags&FlagAppend) != 0 || (attr.Flags&FlagImmutable) != 0 {
        return syscall.EPERM
}
```

**Comment by zhijian-pro** [comment]
`doRename` should check flags
```
if (sattr.Flags&FlagAppend) != 0 || (sattr.Flags&FlagImmutable) != 0 || (dattr.Flags&FlagImmutable) != 0 || (iattr.Flags&FlagAppend) != 0 || (iattr.Flags&FlagImmutable) != 0 {
	return syscall.EPERM
}
```

**Comment by SandyXSD** [comment]
@qingyunha Do you have time to continue this PR? If not, we can take is over.

**Comment by qingyunha** [comment]
Sorry for the delay. I still want to try to implement this feature, perhaps this weekend.

**Comment by SandyXSD** [comment]
Sure

File: pkg/vfs/vfs_unix.go [hunk 5]
```diff
@@ -290,3 +291,27 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
...
 	return
 }
+
+const (
+	FS_IOC_GETFLAGS = 0x80086601
+	FS_IOC_SETFLAGS = 0x40086602
+)
+
+func (v *VFS) Ioctl(ctx Context, ino Ino, cmd uint32, arg uint64, data []byte) (err syscall.Errno) {
+	if (cmd != FS_IOC_GETFLAGS && cmd != FS_IOC_SETFLAGS) || len(data) != 8 {
```
Comment (SandyXSD) [inline pkg/vfs/vfs_unix.go:301]: The length of data is 4 in Ubuntu 20.04

File: pkg/vfs/vfs_unix.go [hunk 6]
```diff
@@ -290,3 +291,27 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
 	}
 	return
 }
+
+const (
+	FS_IOC_GETFLAGS = 0x80086601
+	FS_IOC_SETFLAGS = 0x40086602
+)
+
+func (v *VFS) Ioctl(ctx Context, ino Ino, cmd uint32, arg uint64, data []byte) (err syscall.Errno) {
```
Comment (SandyXSD) [inline pkg/vfs/vfs_unix.go:300]: add `logit` function for this

**Comment by qingyunha** [comment]
Updated.

File: go.mod [hunk 7]
```diff
@@ -231,7 +231,7 @@ require (
 
 replace github.com/minio/minio v0.0.0-20210206053228-97fe57bba92c => github.com/juicedata/minio v0.0.0-20220613143934-cee0571a1b03
 
-replace github.com/hanwen/go-fuse/v2 v2.1.1-0.20210611132105-24a1dfe6b4f8 => github.com/juicedata/go-fuse/v2 v2.1.1-0.20210926080226-cfe1ec802a7f
+replace github.com/hanwen/go-fuse/v2 v2.1.1-0.20210611132105-24a1dfe6b4f8 => github.com/qingyunha/go-fuse/v2 v2.1.1-0.20220710020838-753b234a08c0
```
Comment (davies) [inline go.mod:234]: please send a PR to github.com/juicedata/go-fuse/v2, we should review the changes there too.

File: pkg/vfs/vfs_unix.go [hunk 8]
```diff
@@ -290,3 +291,30 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
...
+		return
+	}
+	var attr = &Attr{}
+	if cmd == FS_IOC_GETFLAGS {
+		err = v.Meta.GetAttr(ctx, ino, attr)
+		utils.NativeEndian.PutUint64(data, uint64(attr.Flags))
+		return
+	}
+	iflag := utils.NativeEndian.Uint64(data)
+	attr.Flags = uint8(iflag)
```
Comment (davies) [inline pkg/vfs/vfs_unix.go:317]: There could be some flags greater than 256, and these flags could be different across OS, so we need to mapping these flags into internal ones (less than 256).

File: go.mod (hunk 7 above)
Comment (qingyunha) [reply go.mod:234]: There is the RP  https://github.com/juicedata/go-fuse/pull/2 .

File: pkg/vfs/vfs_unix.go (hunk 8 above)
Comment (qingyunha) [reply pkg/vfs/vfs_unix.go:317]: Juicefs appears to support Linux, Windows, and MacOS. I can't find out how Windows and MacOS do these things. For MacOS, there is a chflags command, but I don't have a MaC to test it :smiley: . And the FS_IOC_GETFLAGS/FS_IOC_SETFLAGS seems to be Linux-specific.

File: pkg/vfs/vfs_unix.go (hunk 5 above)
Comment (qingyunha) [reply pkg/vfs/vfs_unix.go]: https://elixir.bootlin.com/linux/v5.19.2/source/include/uapi/linux/fs.h#L205

File: pkg/vfs/vfs_unix.go (hunk 8 above)
Comment (davies) [reply pkg/vfs/vfs_unix.go:317]: We may want to support these flags in macOS as well, see http://hints.macworld.com/article.php?story=20031017061722471

Btw, even in Linux, the flags can be wider than 8bits, we should do the mapping (from 32/64bits to 8bits).

So, so we can start with the ioctl callback, then add some flags one by one.

Comment (qingyunha) [reply pkg/vfs/vfs_unix.go:317]: > Btw, even in Linux, the flags can be wider than 8bits, we should do the mapping (from 32/64bits to 8bits).
>
The flags is defined [here](https://elixir.bootlin.com/linux/v5.19.2/source/include/uapi/linux/fs.h#L239), the flags we want to support is fit in low 8 bit, do we need the mapping.
 
> So, so we can start with the ioctl callback, then add some flags one by one.
>
Sorry, I don't get the point.

Comment (davies) [reply pkg/vfs/vfs_unix.go:317]: We may want to support other flags which is greater than 256, so it's better to have the mapping.

Comment (qingyunha) [reply pkg/vfs/vfs_unix.go]: Done the mapping, please take a look.

File: pkg/vfs/vfs_unix.go [hunk 9]
```diff
@@ -290,3 +291,66 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
 	}
 	return
 }
+
+func (v *VFS) Ioctl(ctx Context, ino Ino, cmd uint32, arg uint64, data []byte) (err syscall.Errno) {
```
Comment (SandyXSD) [inline pkg/vfs/vfs_unix.go:295]: `arg` is not used

Comment (qingyunha) [reply pkg/vfs/vfs_unix.go:295]: Maybe other cmd will use the arg directly?

File: pkg/vfs/vfs_unix.go [hunk 10]
```diff
@@ -290,3 +291,66 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
...
+	case FS_IOC_SETFLAGS_32:
+		getiflag = func() uint64 {
+			return uint64(utils.NativeEndian.Uint32(data))
+		}
+	case FS_IOC_GETFLAGS_32:
+		setiflag = func(flag uint64) {
+			utils.NativeEndian.PutUint32(data, uint32(flag))
+		}
+	}
+	if IsSpecialNode(ino) {
```
Comment (SandyXSD) [inline pkg/vfs/vfs_unix.go:330]: this can be checked earlier

File: pkg/vfs/vfs_unix.go [hunk 11]
```diff
@@ -290,3 +291,66 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
...
+	default:
+		err = syscall.ENOTTY
+		return
+	case FS_IOC_SETFLAGS:
+		getiflag = func() uint64 {
+			return utils.NativeEndian.Uint64(data)
+		}
+	case FS_IOC_GETFLAGS:
+		setiflag = func(flag uint64) {
+			utils.NativeEndian.PutUint64(data, flag)
```
[the remark this answers is not in the document]
Comment (qingyunha) [reply pkg/vfs/vfs_unix.go:319]: How do you trigger the ioctl syscall, `lsattr` ?
I don't have a Ubuntu20 in hand, but in Centos7 and Arch, both are 64 bits.
As the last field `Outsize` of `_IoctlIn` show:
```
2022-08-22 15:36:58.469182 I | rx 81: IOCTL n2 *fuse._IoctlIn: &{{72 39 81 2 {{0 0} 5979} 0} 12 0 2148034049 140737010950204 0 8}
2022-08-22 15:36:58.469221 I | tx 81:     38=function not implemented

```

[the remark this answers is not in the document]
Comment (SandyXSD) [reply pkg/vfs/vfs_unix.go:319]: Yes, `lsattr`, and `chattr`

[the remark this answers is not in the document]
Comment (SandyXSD) [reply pkg/vfs/vfs_unix.go:319]:
```
2022-08-22 16:37:56.461336 I | rx 28: IOCTL n402 *fuse.IoctlIn: &{{72 39 28 402 {{0 0} 2843628} 0} 2 0 2148034049 140732591905964 0 4}
panic: runtime error: index out of range [7] with length 4

goroutine 1 [running]:
encoding/binary.littleEndian.PutUint64(...)
        /usr/local/go/src/encoding/binary/binary.go:83
github.com/juicedata/juicefs/pkg/vfs.(*VFS).Ioctl.func3(0xc00006b400?)
        /root/sandy/juicefs/pkg/vfs/vfs_unix.go:319 +0x3b
github.com/juicedata/juicefs/pkg/vfs.(*VFS).Ioctl(0xc0008766e0, {0x309ffc8?, 0xc00099a180}, 0x180?, 0x80086601, 0x3089a3e09?, {0xc000836118?, 0x0?, 0xc00178f4b8?})
        /root/sandy/juicefs/pkg/vfs/vfs_unix.go:353 +0x3c2
github.com/juicedata/juicefs/pkg/fuse.(*fileSystem).Ioctl(0xc00082a460, 0xc00178f4a0?, 0xc000836198, {0xc000836118, 0x4, 0x80})
        /root/sandy/juicefs/pkg/fuse/fuse.go:433 +0xd4
github.com/hanwen/go-fuse/v2/fuse.doIoctl(0xc00178f558?, 0xc000836000)
        /var/lib/go/pkg/mod/github.com/qingyunha/go-fuse/v2@v2.1.1-0.20220710020838-753b234a08c0/fuse/opcode.go:467 +0x13a
github.com/hanwen/go-fuse/v2/fuse.(*Server).handleRequest(0xc0006ff600, 0xc000836000)
```

[the remark this answers is not in the document]
Comment (qingyunha) [reply pkg/vfs/vfs_unix.go:319]: Wird. The ize should be determined using _IOC SIZE (cmd).  https://elixir.bootlin.com/linux/v5.4/source/fs/fuse/file.c#L2756

[the remark this answers is not in the document]
Comment (SandyXSD) [reply pkg/vfs/vfs_unix.go:319]: you may pick the func based on the in.{In,Out}Size instead of cmd.

[the remark this answers is not in the document]
Comment (qingyunha) [reply pkg/vfs/vfs_unix.go]: Updated.

File: pkg/fuse/fuse.go [hunk 12]
```diff
@@ -427,6 +427,13 @@ func (fs *fileSystem) StatFs(cancel <-chan struct{}, in *fuse.InHeader, out *fus
 	return 0
 }
 
+func (fs *fileSystem) Ioctl(cancel <-chan struct{}, in *fuse.IoctlIn, out *fuse.IoctlOut, bufIn, bufOut []byte) (status fuse.Status) {
+	ctx := newContext(cancel, &in.InHeader)
+	defer releaseContext(ctx)
+	out.Result = int32(fs.v.Ioctl(ctx, Ino(in.NodeId), in.Cmd, in.Arg, bufIn, bufOut))
```
Comment (SandyXSD) [inline pkg/fuse/fuse.go:433]:
```suggestion
	return fuse.Status(fs.v.Ioctl(ctx, Ino(in.NodeId), in.Cmd, in.Arg, bufIn, bufOut))
```

File: pkg/vfs/vfs_unix.go [hunk 13]
```diff
@@ -290,3 +291,64 @@ func (v *VFS) Flock(ctx Context, ino Ino, fh uint64, owner uint64, typ uint32, b
...
+		err = syscall.EPERM
+		return
+	}
+	var attr = &Attr{}
+	if cmd>>30 == 1 { // set
+		var iflag uint64
+		if len(bufIn) == 8 {
+			iflag = utils.NativeEndian.Uint64(bufIn)
+		} else if len(bufIn) == 4 {
+			iflag = uint64(utils.NativeEndian.Uint32(bufOut))
```
Comment (SandyXSD) [inline pkg/vfs/vfs_unix.go:323]:
```suggestion
			iflag = uint64(utils.NativeEndian.Uint32(bufIn))
```

**Comment by SandyXSD** [comment]
@qingyunha Please rebase the PR and resolve the conflicts.

File: pkg/fuse/fuse.go (hunk 12 above)
Comment (qingyunha) [reply pkg/fuse/fuse.go:433]: I think the `ioctl` result should be set to out.Result. The return error will be the fuse protocol level error.

**Comment by SandyXSD** [comment]
Failed test:
```
 base_test.go:1570: fallocate f : operation not permitted
```

**Comment by qingyunha** [comment]
> Failed test:
> 
> ```
>  base_test.go:1570: fallocate f : operation not permitted
> ```

Done.

**Review by SandyXSD — APPROVED** [review]

**Pull request merged**